### PR TITLE
[Feature] Neighbor-hood based sampling APIs

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

Implemented the following APIs from RFC #1199 .
- [x] `dgl.sample_neighbors`
- [x] `dgl.sample_neighbors_topk`
- [x] `dgl.in_subgraph`
- [x] `dgl.out_subgraph`

~One question is shall these APIs calculate ID mappings? Will implement that if needed. @BarclayII~ The answer is yes. 

Several fixes on C++ APIs for creating heterographs.
* Added `CreateFromCOO` and `CreateFromCSR` into header file.
* UnitGraph is no longer exposed to python. All APIs returning heterograph object returns a shared pointer of HeteroGraph, **even if there is only one type of edges**.
* For C++ API developer, feel free to use either `CreateFromCOO` or `UnitGraph::CreateFromCOO` when constructing each relation graph. `CreateHeteroGraph` will make sure that the `relation_graphs_` vector stores only `UnitGraph` pointers.

Other usability change.
* The data arrays of `CSRMatrix` and `COOMatrix` are restricted to be IdArrays. Floating type data arrays are not allowed, which makes these matrices for index purposes alone. As a result, `ATEN_CSR_IDX_SWITCH` and `ATEN_COO_IDX_SWITCH` are removed.
* `CSRMatrix` now allows data array to be empty.
* Add `row_sorted` and `col_sorted` flags to COOMatrix.
* Did a sweep on all the CSR/COO operators to make sure that
  - They work on empty data array.
  - Properly handle the sorted flags.
* Expose four interfaces in `BaseHeteroGraph`:
  - [x] GetCOOMatrix
  - [x] GetCSRMatrix
  - [x] GetCSCMatrix
  - [x] SelectFormat

Ready for review but yet some TODOs:
- [x] Test sampling out edges.
- [ ] More doc.
- [x] Return induced edge IDs.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: src/graph/sampling/neighbor/neighbor_cpu.h [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    subrels[etype] = UnitGraph::CreateFromCOO(
+      hg->GetRelationGraph(etype)->NumVertexTypes(),
+      hg->NumVertices(src_vtype),
+      hg->NumVertices(dst_vtype),
+      row, col);
+  }
+
+  return CreateHeteroGraph(hg->meta_graph(), subrels);
+}
```
Comment (BarclayII) [inline src/graph/sampling/neighbor/neighbor_cpu.h]: I'm thinking of representing  neighbor sampling as a sparse matrix operation, like "for each row, randomly take K non-zero entries".

Also, we probably need a COO-based implementation as well.

Comment (jermainewang) [reply src/graph/sampling/neighbor/neighbor_cpu.h]: That's a good idea. Will incorporate your changes.

File: include/dgl/random.h [hunk 2]
```diff
@@ -92,18 +93,51 @@ class RandomEngine {
...
+   * \param prob Array of N unnormalized probability of each element.  Must be non-negative.
+   * \return An integer randomly picked from 0 to N-1.
    */
   template<typename IdxType>
   IdxType Choice(FloatArray prob);
 
+  /*!
+   * \brief Pick random integers between 0 to N-1 according to given probabilities
+   * 
+   * If replace is false, the number of picked integers must not larger than N.
```
Comment (BarclayII) [inline include/dgl/random.h]: "must not be larger than N"

File: include/dgl/base_heterograph.h [hunk 3]
```diff
@@ -508,6 +607,24 @@ HeteroGraphPtr CreateHeteroGraph(
...
+HeteroGraphPtr InEdgeGraph(const HeteroGraphPtr graph, const std::vector<IdArray>& nodes);
+
+/*!
+ * \brief Extract the subgraph of the out edges of the given nodes.
+ * \param graph Graph
+ * \param nodes Node IDs of each type
+ * \return Graph containing only the out edges. The returned graph has the same
+ *         schema as the original one.
+ */
+HeteroGraphPtr OutEdgeGraph(const HeteroGraphPtr graph, const std::vector<IdArray>& nodes);
```
Comment (BarclayII) [inline include/dgl/base_heterograph.h]: Make these also `HeteroSubgraph`s?

File: src/graph/heterograph.cc [hunk 4]
```diff
@@ -218,6 +218,17 @@ HeteroGraph::HeteroGraph(GraphPtr meta_graph, const std::vector<HeteroGraphPtr>&
...
   }
+
+  relation_graphs_.resize(rel_graphs.size());
+  for (size_t i = 0; i < rel_graphs.size(); ++i) {
+    HeteroGraphPtr relg = rel_graphs[i];
+    if (std::dynamic_pointer_cast<UnitGraph>(relg)) {
+      relation_graphs_[i] = std::dynamic_pointer_cast<UnitGraph>(relg);
+    } else {
+      relation_graphs_[i] = CHECK_NOTNULL(
+          std::dynamic_pointer_cast<UnitGraph>(relg->GetRelationGraph(0)));
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/graph/heterograph.cc]: Yes, that is allowed. In fact, `hetero_from_relations` is a restricted version of `union` operator -- requiring all the input graphs to have only one distinct relation.

File: include/dgl/base_heterograph.h [hunk 5]
```diff
@@ -323,6 +341,8 @@ class BaseHeteroGraph : public runtime::Object {
   /*!
    * \brief Get the adjacency matrix of the graph.
    *
+   * TODO(minjie): deprecate this interface; replace it with GetXXXMatrix.
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply include/dgl/base_heterograph.h]: +1 for directly remove. It won't break user codes in python anyway.

**Comment by jermainewang** [comment]
Documentation has not yet been fully finished but will merge this first. Other remaining tasks:
- [ ] More tests for the CSR/CSC operator changes (mainly testing missing data array and integrity of sorted/unsorted)
- [ ] Tests for ascending/descending topk

**Pull request merged**